### ISSUE TITLE
vips_colourspace: no known route between 'cmyk' and 'srgb'

### ISSUE THREAD

**Issue by ferryfax** [jcupitt/libvips#152]
Quick question. To avoid this message from vipsthumbnail do I need to supply an appropriate icc profile?

**Comment by tomasc** [comment]
First, @jcupitt, congratulations on the fantastic library!

I am running into the `vips_colourspace: no known route between 'cmyk' and 'srgb'` issue with `libvips 8.3.3` on debian 8.5.

I am trying to convert a page from a PDF to JPG. The conversion works fine with the same file and same `libvips` version on OS X.

The exact command:

```
vipsthumbnail my.pdf[page=5] -o out.jpg -s 595x842
```

The errors:

```
vipsthumbnail: unable to thumbnail my.pdf[page=5]
vips_colourspace: no known route between 'cmyk' and 'srgb'
```

Is it possible I am missing a library handling this conversion on the server? (I checked and I do have both `liblcms2-2:amd64` as well as `liblcms2-dev:amd64` installed.)

Thank you for your help.

**Comment by jcupitt** [comment]
Hi @tomasc, libvips now has two ways to import PDF files, I think your two installs are using different paths. 

The old way is via libMagick. This uses (in turn) `gs` to import CMYK PDFs as true CMYK images, so you'll need to get a profile from somewhere if you want to convert to sRGB for save.

The new way is via libpoppler and the `pdfload` operation. This is much, much faster and more memory efficient, but it always generates an sRGB image, it cannot make CMYK. 

Unless you really need exact CMYK values, I would suggest the new libpoppler route. I would check for `poppler-glib` on the linux server.

**Comment by tomasc** [comment]
hi @jcupitt thanks a lot for getting back to me so quickly. i saw your posts re poppler & pdf so i took care to install `libpoppler-glib-dev`, `libpoppler-glib8` etc. and indeed i have `libmagic` installed as well. could it be that the `libmagic` takes precedence? is there a way to enforce the use of `poppler` or do i simply need to uninstall everythig related to `libmagic`?

**Comment by jcupitt** [comment]
libMagick will always come last. Check the output of configure and make sure you have something like:

```
PDF import with poppler-glib: 		yes
  (requires poppler-glib 0.16.0 or later)
```

near the end. You can verify that poppler really is being used to read your PDFs with `vipsheader`. You should see:

```
$ vipsheader nipguide.pdf 
nipguide.pdf: 595x841 uchar, 4 bands, srgb, pdfload
```

ie. the libpoppler `pdfload` operation was used to load the image.

**Comment by tomasc** [comment]
@jcupitt you're right!

```
595x842 ushort, 4 bands, cmyk, magickload
```

I will try to recompile, hope that solves the issue.

**Comment by jcupitt** [comment]
Yes, you must reconfigure vips after installing `libpoppler-glib-dev`.

**Comment by tomasc** [comment]
@jcupitt thanks again for help, i compiled libpoppler from source (there is no >= 0.3 version package for debian jessie for example), recompiled vips and eveything now works perfectly.

**Comment by jcupitt** [comment]
Good to hear! 

I noticed your dragonfly with vips, it looks nice. Did you consider using ruby-vips rather than a command-line interface? It should be possible to get a useful speedup by calling the library directly.

**Comment by tomasc** [comment]
Interesting!

What the dragonfly gems do is that they chain processing of files -- they convert one tempfile into another. Intuitively I thought command line would be fastest for this (esp. knowing ruby is not particularly known for its performance). Besides wouldn't I have to reimplement all the nice work you did for vipsthumbnail? Why do you think that using the ruby-vips would be more performant?

**Comment by jcupitt** [comment]
It depends what you are doing. For thumbnailing, ruby-vips would be no faster, but if you are doing more complex processing, like watermarking or adding a drop shadow, ruby-vips will be a lot quicker. 

The command-line interface runs operations one after the other, with a lengthy read phase and write phase between each one as images are written to temp files and then immediately read back in. 

ruby-vips can use the libvips chaining mechanism. If you run a sequence of operations in ruby-vips, perhaps:

```ruby
im = Vips::Image.new_from_file "test.tif", :access => :sequential
text = Vips::Image.text "Hello world!", :dpi => 300
text = (text * 0.3).cast(:uchar)
text = text.embed 100, 100, text.width + 200, text.height + 200
text = text.replicate 1 + im.width / text.width, 1 + im.height / text.height
text = text.crop 0, 0, im.width, im.height
im = text.ifthenelse [255, 0, 0], im, :blend => true
im.write_to_file "out.tif"
```

(the watermarking example), ruby-vips will not actually execute any of the operations until you reach the final `.write_to_file`, it just constructs an image-processing pipeline. When you execute the final write, the whole pipeline runs at the same time, all in parallel. This gives a huge speedup on systems with many cores. Moreover, there's no load/save between operations, which gives another big speedup. And finally the images are streamed, ie. they pass through the pipeline in small 128x128 pixel chunks, so there's an enormous memory saving too.

I'll try a command-line version of that thing and time it, hang on.

**Comment by jcupitt** [comment]
Here's watermarking in sh and ruby-vips:

```ruby
#!/usr/bin/ruby

require 'vips'

im = Vips::Image.new_from_file ARGV[0], :access => :sequential
text = Vips::Image.text ARGV[2], :dpi => 300
text = (text * 0.3).cast(:uchar)
text = text.embed 100, 100, text.width + 200, text.height + 200
text = text.replicate 1 + im.width / text.width, 1 + im.height / text.height
text = text.crop 0, 0, im.width, im.height
im = text.ifthenelse [255, 0, 0], im, :blend => true
im.write_to_file ARGV[1]
```

```bash
#!/bin/bash

infile=$1
outfile=$2
str="$3"

# make text mask
vips text t1.v "$str" --dpi 300
vips linear t1.v t2.v 0.3 0
vips cast t2.v t1.v uchar

text_width=$(vipsheader -f width t1.v)
text_height=$(vipsheader -f height t1.v)

text_width=$((text_width + 200))
text_height=$((text_height + 200))
vips embed t1.v t2.v 100 100 $text_width $text_height

im_width=$(vipsheader -f width $infile)
im_height=$(vipsheader -f height $infile)

across=$((1 + im_width / text_width))
down=$((1 + im_height / text_height))

vips replicate t2.v t1.v $across $down
vips crop t1.v text.v 0 0 $im_width $im_height

# make solid colour
vips black t1.v 1 1
vips linear t1.v t2.v 1 "255 0 0"
vips cast t2.v t1.v uchar
vips embed t1.v colour.v 0 0 $im_width $im_height --extend copy

# combine
vips ifthenelse text.v colour.v $infile $outfile --blend
```

With a large image on my little two-core laptop, I see:

```
$ time ./watermark.rb ~/pics/wtc.jpg x.tif "hello world"
real	0m2.006s
user	0m3.852s
sys	0m0.280s
$ time ./watermark.sh ~/pics/wtc.jpg x.tif "hello world"
real	0m3.867s
user	0m4.384s
sys	0m1.340s
```

So about a 2x speedup. It would be more on a larger machine, I expect.

**Comment by jcupitt** [comment]
Last post: of course you could keep shelling out to run `vipsthumbnail`, but implement other stuff in ruby-vips, perhaps.

**Comment by tomasc** [comment]
I see, that's very good, solid performance gain. I will def. switch over to ruby for some more complex operations. Thanks a lot for the explanation.

BTW I thought it would be very nice to have some of the image magick's geometry behaviour implemented in the `vipsthumbnail` – for example ability to specify width or height only (`NNx` and `xNN`) and no resize when source is larger or smaller (`NNxNN>` and `NNxNN<`) – see http://www.imagemagick.org/script/command-line-processing.php#geometry

Would you be open for adding these? I wrote the necessary calculations here in Ruby https://github.com/tomasc/dragonfly_libvips/blob/master/lib/dragonfly_libvips/dimensions.rb and would like to help to get them included `vipsthumbnail` as these seem to be frequently used when generating image thumbnails.

**Comment by jcupitt** [comment]
Sure, why not, open a new issue and tag it as an enhancement.

**Comment by tomasc** [comment]
Great, I posted as #503 but can't seem to be able to add labels. Let me know if I can help implementing this.

**Comment by andrews05** [comment]
Run into this problem myself for a cmyk image with no profile. Is the some way I can tell it to use a default profile or something? I mean, if I don't know in advance whether the image has a profile or not, how can I provide a default profile without overriding any existing one?

**Comment by jcupitt** [comment]
Hello @andrews05, sure, you can give a fallback profile to `thumbnail`:

http://jcupitt.github.io/libvips/API/current/libvips-resample.html#vips-thumbnail

From the command-line, for example:

```
$ vips thumbnail cmyk.tif x.jpg 128 --import-profile generic-cmyk.icm
```

And it'll use the generic profile you specify if the TIFF file does not have oine embedded.

**Comment by jcupitt** [comment]
In case anyone stumbles across this old issue -- thanks to work by tomasc, since libvips 8.5 you can give geometry specifications like `128x>` or '99x99!` to `vipsthumbnail`.

**Comment by andrews05** [comment]
Thank you. Is it possible to use thumbnail without changing the colorspace? Just keep a cmyk image as cmyk, without having to specify any profiles.

**Comment by jcupitt** [comment]
Hello again, it does the shrinking in RGB colourspace, so CMYK images are actually processed as CMYK -> RGB -> CMYK. This is because CMYK is very non-linear and simply averaging pixels gives ugly results.

You're right, to get the behaviour you want at the moment you need a separate command for CMYK images. 

I'll have a think, perhaps this can be improved.

**Comment by jcupitt** [comment]
I think I found an improvement! `icc_import` was not attaching the fallback profile to the output image if it used it, so `thumbnail` was unable to get back to CMYK. I've added a thing to 8.6 to fix this and I now see:

```
john@kiwi:~/pics$ vipsthumbnail cmyk-no-profile.jpg 
vipsthumbnail: unable to thumbnail cmyk-no-profile.jpg
vips_colourspace: no known route from 'cmyk' to 'srgb'
john@kiwi:~/pics$ vipsthumbnail cmyk-no-profile.jpg --iprofile ~/GIT/nip2/share/nip2/data/cmyk.icm 
john@kiwi:~/pics$ vipsheader tn_cmyk-no-profile.jpg 
tn_cmyk-no-profile.jpg: 128x85 uchar, 4 bands, cmyk, jpegload
```

So it's picking up the fallback profile to get back to CMYK again after the thumbnail operation.

This change will be in 8.6.3, due in a few days. Thanks!

**Comment by andrews05** [comment]
Sounds great, that should be a big help.

**Comment by gertcuykens** [comment]
Testfile: https://drive.google.com/file/d/1b7FzZZKJhrLnlb-5bYsUwuzRSEhKltBA/view?usp=sharing
Version: vips-8.6.4-Wed Jun 13 15:24:52 UTC 2018

vips colourspace test_CMYK.jpg rgb.jpeg srgb
vips_colourspace: no known route from 'cmyk' to 'srgb'

Did I do something wrong to convert to srgb or rgb?

**Comment by jcupitt** [comment]
Hello, convert CMYK jpg images to sRGB like this:

```
$ vips icc_transform cmyk.jpg x.jpg /usr/share/color/icc/colord/sRGB.icc 
```

That's assuming your jpg image has an embedded CMYK profile. 

libvips does not have a built-in CMYK -> $x converter, you need to use lcms via the `icc_import` / `icc_export` / `icc_transform` operations.

**Comment by gertcuykens** [comment]
Ok thx both cli's are working
```
vips icc_transform test_CMYK.jpg rgb.jpg ../icc/RGB/AppleRGB.icc
```
and 
```
vips icc_import test_CMYK.jpg rgb.jpg --input-profile icc/CMYK/USWebCoatedSWOP.icc
```
icc_import saves it to the color profile selected but also to color space RGB, i expect it would have kept the CMYK color space?

**Comment by jcupitt** [comment]
`icc_import` takes the image to PCS (CIELAB by default). When you write a CIELAB image to JPG, libvips uses an internal converter to make sRGB for you. Try:

```
john@yingna ~/pics $ vips icc_import cmyk.jpg x.v
john@yingna ~/pics $ vipsheader -a x.v
x.v: 1419x1001 float, 3 bands, lab, jpegload
width: 1419
height: 1001
bands: 3
format: float
coding: none
interpretation: lab
xoffset: 0
yoffset: 0
xres: 11.808
yres: 11.808
filename: x.v
vips-loader: jpegload
jpeg-multiscan: 0
jpeg-chroma-subsample: 4:4:4:4
exif-data: 186 bytes of binary data
resolution-unit: in
...
icc-profile-data: 961644 bytes of binary data
```

So the vips format image still has the CMYK profile. It's the implicit conversion on jpg write that drops it.

**Comment by gertcuykens** [comment]
Thx ps is their a way to get the vipsheader output when using the govips that you can tell at first look ?

https://godoc.org/github.com/davidbyttow/govips/pkg/vips

Or do I need to find a other go wrap library specifically for vipsheader?

**Comment by jcupitt** [comment]
Sorry, I don't know, I didn't make govips, you would need to ask there. In Python, for example, it's `image.get('icc-profile-data')`. Maybe something similar?

Please open a new issue if you have a new question: it'll make it easier for others to find answers.

**Comment by jcupitt** [comment]
Please don't post on closed issues, it just causes confusion. 

It's not really a bug -- `colourspace` just can't convert to CMYK. Use `icc_transform` plus a profile instead.

**Comment by adityapatadia** [comment]
Sorry I missed your last comment. I will create new one if I need further information.
